ocaml/test/test_event.ml

```diff
@@ -0,0 +1,286 @@
+         with e ->
+           Printf.printf "Caught unexpected error: %s\n" (ExnHelper.string_of_exn e);
+           ok := false;
+      ) () in
+  let (interfering_thread: Thread.t) = Thread.create
+      (fun () ->
+         wait_for_pool_key __context key
+      ) () in
+  Thread.delay 0.5; (* wait for both threads to block in Event.from *)
+  Db.Pool.add_to_other_config __context pool key "1";
+  Thread.join interfering_thread;
+  Thread.join i_should_succeed;
+  (* Check that Event.from didn't get cancelled by mistake *)
+  assert_bool "event_from_parallel_test" !ok
+
+let object_level_event_test session_id =
+  let __context, session_id = event_setup_common () in
+  let m = Mutex.create () in
+  let finished = ref false in
+  let reported_failure = ref false in
+  let vm_a = make_vm ~__context ~name_label:"vm_a" () in
+  let vm_b = make_vm ~__context ~name_label:"vm_b" () in
+
+  Printf.printf "watching %s\n%!" (Ref.string_of vm_a);
+  Printf.printf "ignoring %s\n%!" (Ref.string_of vm_b);
+
+  let key = "object_level_event_next" in
+
+  begin try Db.VM.remove_from_other_config __context vm_a key with _ -> () end;
+  begin try Db.VM.remove_from_other_config __context vm_b key with _ -> () end;
+
+  let (_: Thread.t) = Thread.create
+      (fun () ->
+         let token = ref "" in
+         while not (Mutex.execute m (fun () -> !finished)) do
+           Printf.printf "Calling event.from...\n%!";
+           let events = Xapi_event.from __context [ Printf.sprintf "vm/%s" (Ref.string_of vm_a) ] (!token) 10. |> parse_event_from in
+           Printf.printf "Got %d events\n%!" (List.length events.events);
+           List.iter
+             (fun event ->
+                if event.reference <> Ref.string_of vm_a then begin
+                  Printf.printf "event on %s which we aren't watching\n%!" event.reference;
+                  Mutex.execute m
+                    (fun () ->
+                       reported_failure := true;
+                       finished := true;
+                    )
+                end
+             ) events.events;
+           token := events.token;
+           let oc = Db.VM.get_other_config __context vm_a in
+           if List.mem_assoc key oc && (List.assoc key oc) = "1"
+           then Mutex.execute m (fun () ->
+               Printf.printf "got expected event (new token = %s)\n%!" !token;
+               finished := true;
+             )
+           else Printf.printf "Db doesn't have expected change in...\n%!";
+         done
+      ) () in
+  Thread.delay 0.5;
+  Printf.printf "Adding to vm_b\n%!";
+  Db.VM.add_to_other_config __context vm_b key "1";
+  Thread.delay 0.5;
+  Printf.printf "Removing from vm_b\n%!";
+  Db.VM.remove_from_other_config __context vm_b key;
+  Printf.printf "Adding to vm_a. This ought to wake up the event thread\n%!";
+  Db.VM.add_to_other_config __context vm_a key "1";
+  Thread.delay 1.0;
+  Mutex.execute m
+    (fun () ->
+       if not (!reported_failure) then begin
```

Comment:
Yes, every other call to `failed` in quicktest.ml has a corresponding `assert_bool` in the new test_event.ml file. (That is to say, every one except the one where I commented at line 242-243 of test_event.ml where it's inside the thread so we probably shouldn't use `assert_bool`.)